### ISSUE TITLE
Dispatch opt ideas

### ISSUE THREAD

**Issue by JohnMBrooks** [AcheronVM/acheronvm#3]
Dispatch could be 3 cycles faster by:
1) Moving it to ZP and using abs,y instead of (zp),y addressing
2) Relying on branch/call to avoid iptr offset overflow BMI check

Ex: PLASMA 2.0 FetchOp:
https://github.com/dschmenk/PLASMA/blob/master/src/vmsrc/apple/plvm02.s

**Comment by AcheronVM** [comment]
This dispatcher is larger than others, so I didn't want to sacrifice the ZP space.  Everything inside dispatch.asm would have to be in ZP, not just a few lines.  Code from ops-short and ops-branch also save quite a few bytes and cycles by branching directly back into the dispatcher instead of having to branch and jmp absolute into ZP.

Can you describe point 2 in more detail?  There's a few things that have been tried in other languages, like page-crossing instructions required at the end of pages. In Acheron,  `iptr` is not page-aligned.  Wherever the last branch/jump/call routine pointed to becomes the new `iptr`, and `iptr_offset` is not checked at every increment, but rather only in the dispatcher where it checks for >127.  This allows much shorter and faster instruction implementations, since the operands can be read while incrementing .Y without performing any checking.

**Comment by JohnMBrooks** [comment]
Agreed that putting the dispatcher in ZP is somewhat costly and not always a win.

For more on point 2 and other 6502 interp ideas, check out the PLASMA 2.0 discussion from a year or two ago:

https://groups.google.com/d/msg/comp.sys.apple2.programmer/W5e8Nj8TN5U/VcMHR-MWAwAJ

David Schmenk iterated through several strategies to reduce the cost of avoiding iptr offset overflow, but eventually realized that since the offset is added to iptr on a branch, if there are no more than 255 bytes of non-branch instructions before a branch occurs, then the offset cannot overflow and the check can be omitted. PLASMA has a parser/compiler which generates pcode, so the 255-byte limit could be enforced automatically (by injecting a jump to next op if needed).

Nice work and presentation. Using a register file & window is much more efficient than stack-based expression eval, IMO. Though PLASMA has a very fast stack implementation.

-JB

**Comment by JohnMBrooks** [comment]
Here is the other thread of 6502 VM speed-ups for PLASMA:

https://groups.google.com/d/msg/comp.sys.apple2.programmer/aFY8pKRSaIQ/8A7ofhYaAQAJ

-JB
